app/services/unpublish_handler_service.rb

```diff
@@ -0,0 +1,89 @@
+class UnpublishHandlerService
+  UNPUBLISH_LOG_PATH = "#{Rails.root}/log/unpublish_message.log".freeze
+
+  def initialize(content_id)
+    @content_id = content_id
+  end
+
+  def self.call(*args)
+    new(*args).call
+  end
+
+  def call
+    if email_params.empty?
+      nil
+    elsif taxon_is_unpublished?
+      email_ids = UnpublishEmailBuilder.call(email_params).ids
+      log_taxon_emails(email_ids)
+    else
+      log_none_taxon_lists(subscriber_lists)
+    end
+  end
+
+private
+
+  attr_reader :content_id
+
+  def taxon_is_unpublished?
+    subscriber_lists.first.links.keys.include?(:taxon_tree)
+  end
+
+  def queue_delivery_request_workers(email_ids)
+    email_ids.each do |email_id|
+      DeliveryRequestWorker.perform_async_in_queue(
+        email_id, queue: :delivery_immediate
+      )
+    end
+  end
+
+  def email_params
+    subscriber_lists.flat_map do |subscriber_list|
+      subscriber_list.subscribers.activated.map do |subscriber|
+        {
+          subject: subscriber_list.title,
+          address: subscriber.address,
+          subscriber_id: subscriber.id
+        }
+      end
+    end
+  end
+
+  # For this query to return the content id has to be wrapped in a double quote blame psql 9.3
+  def subscriber_lists
+    @subscriber_lists ||= SubscriberList
+    .where(":id IN (SELECT json_array_elements((json_each(links)).value)::text)", id: "\"#{content_id}\"")
+    .includes(:subscribers).to_a
+  end
+
+  def log_taxon_emails(email_ids)
+    Email.where(id: email_ids).each do |email|
+      logger.info(<<-INFO.strip_heredoc)
+        ----
+        Created Email:
+        id: #{email.id}
+        subject: #{email.subject}
+        body: #{email.body}
+        subscriber_id: #{email.subscriber_id}
+        ----
+      INFO
+    end
+  end
+
+  def log_none_taxon_lists(subscriber_lists)
+    subscriber_lists.each do |list|
+      logger.info(<<-INFO.strip_heredoc)
+        ++++
+        Email not created but matching a SubscriberList (Policy area etc):
```

Comment:
Topics (taxons) use subscriber lists too, maybe say "Not sending notification about non-Topic SubscriberList".

Reply:
👍